format/luajit/luajit.go

```diff
@@ -0,0 +1,409 @@
+	ktabtype := d.FieldULEB128("type", fallbackUintMapSymStr{
+		fallback: "str",
+		UintMapSymStr: scalar.UintMapSymStr{
+			0: "nil",
+			1: "false",
+			2: "true",
+			3: "int",
+			4: "num",
+		},
+	})
+
+	switch ktabtype {
+	case 0:
+		// nil
+		d.FieldValueAny("value", nil)
+
+	case 1:
+		// false
+		d.FieldValueBool("value", false)
+
+	case 2:
+		// true
+		d.FieldValueBool("value", true)
+
+	case 3:
+		// int
+		d.FieldULEB128("value")
+
+	case 4:
+		LuaJITDecodeNum(d)
+
+	// ktabtype >= 5
+	default:
+		// str
+		size := ktabtype - 5
+		d.FieldUTF8("value", int(size))
+	}
+}
+
+func LuaJITDecodeTab(d *decode.D) {
+	narray := d.FieldULEB128("narray")
+	nhash := d.FieldULEB128("nhash")
+
+	d.FieldArray("array", func(d *decode.D) {
+		for i := uint64(0); i < narray; i++ {
+			d.FieldStruct("element", LuaJITDecodeKTabK)
+		}
+	})
+
+	d.FieldArray("hash", func(d *decode.D) {
+		for i := uint64(0); i < nhash; i++ {
+			d.FieldStruct("pair", func(d *decode.D) {
+				d.FieldStruct("key", LuaJITDecodeKTabK)
+				d.FieldStruct("value", LuaJITDecodeKTabK)
+			})
+		}
+	})
+}
+
+func LuaJITDecodeI64(d *decode.D) int64 {
+	lo := d.ULEB128()
+	hi := d.ULEB128()
+	return int64((hi << 32) + lo)
+}
+
+func LuaJITDecodeU64(d *decode.D) uint64 {
+	lo := d.ULEB128()
+	hi := d.ULEB128()
+	return (hi << 32) + lo
+}
+
+func LuaJITDecodeComplex(d *decode.D) {
+	d.FieldAnyFn("real", func(d *decode.D) any {
+		rlo := d.ULEB128()
+		rhi := d.ULEB128()
+		r := (rhi << 32) + rlo
+		return u64tof64(r)
+	})
+
+	d.FieldAnyFn("imag", func(d *decode.D) any {
+		ilo := d.ULEB128()
+		ihi := d.ULEB128()
+		i := (ihi << 32) + ilo
+		return u64tof64(i)
+	})
+}
+
+func LuaJITDecodeKGC(d *decode.D) {
+	kgctype := d.FieldULEB128("type", fallbackUintMapSymStr{
+		fallback: "str",
+		UintMapSymStr: scalar.UintMapSymStr{
+			0: "child",
+			1: "tab",
+			2: "i64",
+			3: "u64",
+			4: "complex",
+		},
+	})
+
+	switch kgctype {
+	case 0:
+		// child
+
+	case 1:
+		LuaJITDecodeTab(d)
+
+	case 2:
+		LuaJITDecodeI64(d)
+
+	case 3:
+		LuaJITDecodeU64(d)
+
+	case 4:
+		// json does not support complex numbers,
+		// so we use a struct{real: float64, imag: float64}
+		d.FieldStruct("value", LuaJITDecodeComplex)
+
+	// kgctype >= 5
+	default:
+		// str
+		size := kgctype - 5
+		d.FieldUTF8("value", int(size))
+	}
+}
+
+func LuaJITDecodeKNum(d *decode.D) any {
+	// knum = intU0 | (loU1 hiU)
+	// ...
+	// W = 32 bit, U = ULEB128 of W, U0/U1 = ULEB128 of W+1
+
+	// intU0 encodes 33 bits : a (signed) int32, plus the LSB=0
+	// loU1 encodes 33 bits : the lower half of a float64, plus the LSB=1
+	// hiU encodes 32 bits : the higher half of the float64
+
+	lo := d.ULEB128()
+	if lo&1 == 0 {
+		// we have an int32 (aka LuaJIT 'int')
+
+		// drop the LSB
+		lo_data := lo >> 1
+
+		// downcast to 32bits. should not overflow
+		lo_data_u32 := uint32(lo_data)
+
+		// make it a signed integer
+		lo_data_i32 := int32(lo_data_u32)
+
+		// return a larger type to make fq happy
```

Comment:
🕺💃  maybe there should be smaller number types hmm, should probably do another round of profiling memory soon